File: app/src/main/java/cah593wang/uwaterloo/cs/student/httpswww/uwcourse/DisplayMessageActivity.java
package cah593wang.uwaterloo.cs.student.httpswww.uwcourse;

import android.os.Handler;
import android.os.Looper;
import android.os.NetworkOnMainThreadException;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ListView;
import android.widget.Toast;

import static android.widget.AdapterView.*;

public class DisplayMessageActivity extends AppCompatActivity {
    public static final String COURSE_PARAMS = "COURSE_PARAMS";

    Course course;
    @Override
    protected void onCreate(Bundle savedInstanceState) throws NetworkOnMainThreadException {

        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_display_message);

        String [] params = getIntent().getStringExtra(COURSE_PARAMS).split(" ");
        course = new Course(params[0], Integer.parseInt(params[1]), Integer.parseInt(params[2])) {
            @Override
            public void onCourseReturned() {

                Handler mainHandler = new Handler(Looper.getMainLooper());
                Runnable myRunnable = new Runnable() {
                    @Override
                    public void run() {
                        initAdapter();
                    } // This is your code
                };
                mainHandler.post(myRunnable);

            }
        };




    }

    private void initAdapter() {
        customListAdapter adapter = new customListAdapter(this, course);
        ListView out = findViewById(R.id.display);
        out.setAdapter(adapter);
        out.setOnItemClickListener(new OnItemClickListener() {
            //TODO add info to database

            @Override
            public void onItemClick(AdapterView<?> adapterView, View view, int i, long l) {
                Toast.makeText(DisplayMessageActivity.this, "todo", Toast.LENGTH_SHORT).show();
            }
        });
        out.setOnItemLongClickListener(new OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
                Toast.makeText(DisplayMessageActivity.this, "todo", Toast.LENGTH_SHORT).show();
                return false;
            }

        });
    }
}

File: app/src/main/java/cah593wang/uwaterloo/cs/student/httpswww/uwcourse/DisplayMessageActivity.kt
package cah593wang.uwaterloo.cs.student.httpswww.uwcourse

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.os.NetworkOnMainThreadException
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.AdapterView.*
import android.widget.Button
import android.widget.ListView
import android.widget.TextView
import android.widget.Toast
import java.lang.Exception

class DisplayMessageActivity : AppCompatActivity() {
    lateinit var course: Course
    lateinit var display: ListView

    @Throws(NetworkOnMainThreadException::class)
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_display_message)

        val department = intent.getStringExtra(DEP)
        val courseNum = intent.getIntExtra(COURSE_NUM, 0)
        val term = intent.getIntExtra(TERM, 0)

        findViewById<TextView>(R.id.availableSections).bringToFront()
        //start the course class to get course info from the webscraper inside the class
            course = object : Course(department, courseNum, term, this.application) {
                override fun onCourseReturned() {
                    val mainHandler = Handler(Looper.getMainLooper());
                    val myRunnable = Runnable {
                        //when the data is returned, use it to initialize the adapter
                        initAdapter()
                    };
                    mainHandler.post(myRunnable)
                }

                override fun onFailed(e: Exception) {
                    finishFail(e.message ?: "ERROR")
                }
            }


        findViewById<Button>(R.id.button).setOnClickListener {
            val returnIntent = Intent()
            val checkBoxes = (display.adapter as CustomListAdapter).checkBoxes
            var serializedResult = ArrayList<Section>()
            for (i in checkBoxes.indices) {
                if (checkBoxes[i]) {
                    serializedResult.add(display.adapter.getItem(i) as Section)
                }
            }
            returnIntent.putExtra("RESULT", serializedResult)
            setResult(Activity.RESULT_OK,returnIntent)
            finish()
        }
    }

    private fun initAdapter() {
        findViewById<TextView>(R.id.loading).visibility = View.GONE

        //initialize the listview adapter with the course info
        val adapter = CustomListAdapter(this, course)
        display = findViewById(R.id.display)
        display.adapter = adapter
    }

    private fun finishFail(result: String) {
        val returnIntent = Intent()
        returnIntent.putExtra("RESULT", result)
        setResult(Activity.RESULT_CANCELED,returnIntent)
        finish()
    }

    companion object {
        const val DEP = "DEP"
        const val COURSE_NUM = "COURSE_NUM"
        const val TERM = "TERM"
    }
}